tests/test_ecu_variant_matching.py

```diff
@@ -0,0 +1,285 @@
+):
+    return DiagLayer(
+        variant_type=DIAG_LAYER_TYPE.ECU_VARIANT,
+        odx_id=OdxLinkId(local_id="ecu_variant3", doc_fragments=doc_frags),
+        short_name="ecu_variant3",
+        services=[identificationService, supplierService],
+        ecu_variant_patterns=[ecu_variant_pattern1, ecu_variant_pattern3],
+        odxlinks=odxlinks,
+    )
+
+
+def as_bytes(dikt):
+    return bytes(json.dumps(dikt), "utf-8")
+
+
+@pytest.mark.parametrize("use_cache", [True, False])
+# the req_resp_mapping maps request to responses for the ecu-under-test
+@pytest.mark.parametrize(
+    "req_resp_mapping, expected_variant",
+    [
+        # test if full match of matching parameters is accepted
+        (
+            {
+                b"\x22\x10\00": as_bytes({"id": 2000}),
+                b"\x22\x20\00": as_bytes({"name": "supplier_B"}),
+            },
+            "ecu_variant2",
+        ),
+        # test if partial match of matching parameters is rejected
+        (
+            {
+                b"\x22\x10\00": as_bytes({"id": 2000}),
+                b"\x22\x20\00": as_bytes({"name": "supplier_C"}),
+            },
+            "ecu_variant3",
+        ),
+        # test if first full match is preferred over second match
+        (
+            {
+                b"\x22\x10\00": as_bytes({"id": 1000}),
+                b"\x22\x20\00": as_bytes({"name": "supplier_A"}),
+            },
+            "ecu_variant1",
+        ),
+    ],
+)
+def test_ecu_variant_matching(
+    ecu_variant_1,
+    ecu_variant_2,
+    ecu_variant_3,
+    use_cache,
+    req_resp_mapping,
+    expected_variant,
+):
+    matcher = EcuVariantMatcher(
+        ecu_variant_candidates=[ecu_variant_1, ecu_variant_2, ecu_variant_3],
+        use_cache=use_cache,
+    )
+    for req in matcher.request_loop():
+        resp = req_resp_mapping[req]
+        matcher.evaluate(resp)
+    assert matcher.has_match()
+    assert matcher.get_active_ecu_variant().short_name == expected_variant
+
+
+@pytest.mark.parametrize("use_cache", [True, False])
+def test_no_match(ecu_variant_1, ecu_variant_2, ecu_variant_3, use_cache):
+    # stores the responses for each request for the ecu-under-test
+    req_resp_mapping = {
+        b"\x22\x10\00": as_bytes({"id": 1000}),
+        b"\x22\x20\00": as_bytes({"name": "supplier_D"}),
+    }
+
+    matcher = EcuVariantMatcher(
+        ecu_variant_candidates=[ecu_variant_1, ecu_variant_2, ecu_variant_3],
+        use_cache=use_cache,
+    )
+    for req in matcher.request_loop():
+        resp = req_resp_mapping[req]
+        matcher.evaluate(resp)
+    assert not matcher.has_match()
+    with pytest.raises(AssertionError):
+        matcher.get_active_ecu_variant()
+
+
+@pytest.mark.parametrize("use_cache", [True, False])
+# test if pending matchers reject the has_match() or active variant query
+def test_no_request_loop(ecu_variant_1, ecu_variant_2, ecu_variant_3, use_cache):
```

Comment:
```suggestion
def test_no_request_loop(ecu_variant_candidates : List[DiagLayer], use_cache: bool) -> None:
```

(same for the other methods with three hardcoded variant arguments.)